### ISSUE TITLE
Feature Request: Removal of jQuery as Dependency

### ISSUE THREAD

**Issue by RacerDelux** [select2/select2#6196]
To start, I know this was asked a while back. The reason I am reposting this feature request is that we are now in 2022 and the use of jQuery simply isn't required in many/most modern frameworks.

While I know this would be a large refactor, the good news is that JavaScript has a 1:1 replacement for pretty much anything you would do in jQuery. And since this project has type definitions, if done in TypeScript it would help quickly find and update all of the jQuery references anyway.

Sadly there are not too many good alternatives to select2 that do not include jQuery. I know many of us would be very glad to see the drop of a dependency.

Apart from the dependency removal, another bonus is that this will remove any bugs that occur from using different jQuery versions.

As a flip side, if this refactor is done, I am not sure if IE11 needs to be supported. Microsoft has dropped support in their own applications (last year I think) and its finally being left in the past. The only reason I mention this is that there are many powerful features that are unlocked the second IE11 is no longer in the picture.

**Comment by ajiho** [comment]
Yes, bootstrap 5 is already doing this, and it has completely abandoned its support for ie11. I personally suggest that select2 should do the same.

**Comment by kevin-brown** [comment]
I would absolutely love to get rid of the dependency on jQuery! But unfortunately there's a lot of ground to cover and only one of me, so I'm willing to accept pull requests tagged to this ticket.

**Comment by RacerDelux** [comment]
> 

I have not looked at source yet, but is this built on TypeScript? If so it would make the switch over process pretty straight forward, just tedious.

If I get some extra time here I will look into doing some PRs. More than likely removing jQuery would be best in steps right? Or would you prefer it be in one PR?

**Comment by kevin-brown** [comment]
Select2 is unfortunately not built on TypeScript (honestly I wish it was), but instead of AMD-compatible JavaScript. And yes, I prefer smaller PRs to make reviews a bit easier and quicker. It also greatly reduces the chances that someone gets half way through it and gives up, making all of their work go to waste.

**Comment by stratboy** [comment]
Yes, jquery should be abandoned here.

**Comment by duanemoody** [comment]
Looking over the 6500 lines of select2.full.js, it's mostly leveraging jQ for events (96 calls to `.on()`) and some property getters/setters, but ironically there's only one complex `$()` query selector in the code with the other 45 wrapping variable names. There is afaik no chaining in the codebase.

There's also a number of older programming habits: iteration still mostly uses length counters, all variable declaration is done with `var` and there's still shimming for versions of IE below 10.  

I honestly think select2 only uses jQuery because the author liked its UI event binding syntax, not because the author had a personal preference for jQ as a whole. 

This could be refactored.

**Comment by kevin-brown** [comment]
> Looking over the 6500 lines of select2.full.js, it's mostly leveraging jQ for events (96 calls to `.on()`) and some property getters/setters, but ironically there's only one complex `$()` query selector in the code with the other 45 wrapping variable names. There is afaik no chaining in the codebase.

I believe there is a fair amount of selector chaining (not action chaining) if you look at `this.$selection` (and similar) references around the code base. But none of it is complex enough that it should cause an issue during a migration, it's all pretty standard and basic stuff. Generally we've tried to minimize the dependency on sizzle and other selector engines.

> There's also a number of older programming habits: iteration still mostly uses length counters, all variable declaration is done with `var` and there's still shimming for versions of IE below 10.  

And we've made the decision in 4.1.x to drop these shims, so that's another win. Ultimately this code base largely dates from 2014ish so it's not a huge surprise that most of the code patterns are that old.

> I honestly think select2 only uses jQuery because the author liked its UI event binding syntax, not because the author had a personal preference for jQ as a whole. 

This is pretty accurate. Over the years we've reduced the reliance on jQuery since it was clearly fading away.

**Comment by stratboy** [comment]
choices.js is a good alternative, I've just tried it, seems well made. Still, I don't think it will be actively developed in future. Therefore, a Select2 porting to ES6 would still be a great thing.

**Comment by duanemoody** [comment]
Once SELECT2 has no dependencies, I'd strongly recommend it be proposed to WHATWYG or similar as the prototype for a native code HTML5 built-in that still gracefully degrades to a regular `SELECT` element for older user agents. That's not in the near future but other initiatives make it clear things are ready for the concept of a more stylable, multicolumn element (and ever since they nerfed the CSS `appearance:` attribute we can't do it in plain CSS).

**Comment by ajiho** [comment]
If we abandon jquery, what should we replace the large number of on methods in jquery in the source code with native JavaScript? As far as I know, this is not an easy task, and the delegation of this event is quite complex. Has anyone worked together to complete this task?

**Comment by duanemoody** [comment]
https://stackoverflow.com/questions/29475295/javascript-equivalent-for-jquery-on-with-additional-selector
This has been asked and answered multiple times because `.on()` is arguably the most powerful feature of jQuery, abstracting binds/listeners away from only existing elements while also improving performance by deduplicating those binds/listeners. 

The code already exists, the only question is how to implement it in a wrapper function with methods that anticipate a future native equivalent to `.on()` and when detected switch to leveraging it instead – just like jQuery itself always has.

**Comment by RacerDelux** [comment]
@duanemoody The most common strategy I use is to use a single javascript event listener on a dropdown that listens to only the scope of that HTML.
(Note: In my example, I have three, but that is because I am listening to three different types of actions)

You put data-attributes like `data-dropdown-action="select"` and `data-dropdown-action="close"` as examples.
The single even listener looks for the `data-dropdown-action` attribute and switches on its value to determine functionality.

A single even listener per dropdown - great performance gains.

I wrote my own dropdown that can do type-ahead, multi or single select, populate a breadcrumb or update URL query parameters, optional placeholder and can be tied to a static or dynamic data source.

Here is a code example from that dropdown:

In my construction (I use TypeScript)
```ts
this.ComponentTarget.addEventListener('mousedown', this.selectAction.bind(this));
this.DropDownInputTarget.addEventListener('blur', this.selectAction.bind(this));
this.DropDownInputTarget.addEventListener('keyup', this.selectAction.bind(this));
```

```ts
private async selectAction(event: Event) {
        const target = event.target as HTMLDivElement;
        if (!target.dataset.dropdownAction) return;

        switch (target.dataset.dropdownAction) {
            case 'search':
                // Search Logic
                break;
            case 'toggle':
                // Open/Close logic
                break;
            case 'item':
                // Logic for when you select an item
                break;
            case 'clearSubset':
                // When you want to clear your selection in a multi-select dropdown
                break;
            case 'ignore':
                event.preventDefault();
                break;
        }
    }
 ```
 Finally, here is an example of the HTML:
 ```html
 <div id="StateMultiSelect" class="custom-input" data-dropdown-component>
     <div class="select-container">
        <input class="form-control" placeholder="State" autocomplete="off"
              data-dropdown-action="search" 
              data-custom-validation="dropdownValidation" 
              data-dropdown-input
              maxlength="50"
              required/>
      <input type="hidden" asp-for="State"/>
      <span class="input-error input-error-inline" aria-live="polite" data-error-value-missing="State Required">
      </span>
      <span class="input-buttons">
          <span class="product-drop-clear" data-dropdown-action="clearSubset">
              <i class="fa-solid fa-times" data-fa-transform="shrink-5"></i>
          </span>
          <span class="product-drop-toggle" data-dropdown-action="toggle">
              <i class="fa-regular fa-chevron-down" data-fa-transform="up-1"></i>
          </span>
      </span>
    </div>
    <div class="select-container-parent">
      <div class="select-data-container items-above custom-scroll">
          <ul>
              <li data-name="No Country Selected" data-dropdown-action="item" data-selected disabled>
                  <span class="item-text">
                      No Country Selected
                  </span>
              </li>
          </ul>
      </div>
    </div>
</div>
 ```

**Comment by ajiho** [comment]
> https://stackoverflow.com/questions/29475295/javascript-equivalent-for-jquery-on-with-additional-selector This has been asked and answered multiple times because `.on()` is arguably the most powerful feature of jQuery, abstracting binds/listeners away from only existing elements while also improving performance by deduplicating those binds/listeners.
> 
> The code already exists, the only question is how to implement it in a wrapper function with methods that anticipate a future native equivalent to `.on()` and when detected switch to leveraging it instead – just like jQuery itself always has.

I have checked the address you provided, but the on solution you provided is not suitable for use in complex production environments. The biggest problem is that its event cannot call stopPropagation. I think we can still use the on method of jQuery. I have tried customizing jQuery that only includes the on method, and it works well. However, the only problem is that even with code compression, its size is close to 30K . If we can tolerate this size, then we can immediately make select2 dependent on jQuery. In the future, we will gradually search for smaller alternatives to the on method

**Comment by ajiho** [comment]
Select2 works very well in non Vue and React projects, and I really like it. Unfortunately, if it doesn't get rid of the dependency on jQuery, the number of people using this library may become fewer and fewer. I call on those who can help to work together to revitalize it. Once Select2 is released without jQuery, I can also use more beautiful document generation tools to write a more beautiful document for Select2, such as Vitepress or Docusaurus

**Comment by RacerDelux** [comment]
> > https://stackoverflow.com/questions/29475295/javascript-equivalent-for-jquery-on-with-additional-selector This has been asked and answered multiple times because `.on()` is arguably the most powerful feature of jQuery, abstracting binds/listeners away from only existing elements while also improving performance by deduplicating those binds/listeners.
> > The code already exists, the only question is how to implement it in a wrapper function with methods that anticipate a future native equivalent to `.on()` and when detected switch to leveraging it instead – just like jQuery itself always has.
> 
> I have checked the address you provided, but the on solution you provided is not suitable for use in complex production environments. The biggest problem is that its event cannot call stopPropagation. I think we can still use the on method of jQuery. I have tried customizing jQuery that only includes the on method, and it works well. However, the only problem is that even with code compression, its size is close to 30K . If we can tolerate this size, then we can immediately make select2 dependent on jQuery. In the future, we will gradually search for smaller alternatives to the on method

jQuery needs to be removed 100%. The "on" method and the way jQuery handles events isn't appropriate for today's web. Javascript can everything jQuery can do, and then some.

Did you get a chance to see the code I provided above?

**Comment by ajiho** [comment]
> > > https://stackoverflow.com/questions/29475295/javascript-equivalent-for-jquery-on-with-additional-selector This has been asked and answered multiple times because `.on()` is arguably the most powerful feature of jQuery, abstracting binds/listeners away from only existing elements while also improving performance by deduplicating those binds/listeners.
> > > The code already exists, the only question is how to implement it in a wrapper function with methods that anticipate a future native equivalent to `.on()` and when detected switch to leveraging it instead – just like jQuery itself always has.
> > 
> > 
> > I have checked the address you provided, but the on solution you provided is not suitable for use in complex production environments. The biggest problem is that its event cannot call stopPropagation. I think we can still use the on method of jQuery. I have tried customizing jQuery that only includes the on method, and it works well. However, the only problem is that even with code compression, its size is close to 30K . If we can tolerate this size, then we can immediately make select2 dependent on jQuery. In the future, we will gradually search for smaller alternatives to the on method
> 
> jQuery needs to be removed 100%. The "on" method and the way jQuery handles events isn't appropriate for today's web. Javascript can everything jQuery can do, and then some.
> 
> Did you get a chance to see the code I provided above?

Yes, I completely agree with your point of view. I have reviewed them and they are too simple to be used in production environments. They are not a truly effective modern alternative to jQuery's on method. I like select2. It should be migrated to ESM and moved away from jQuery as soon as possible. We need to work together

**Comment by ajiho** [comment]
> > > https://stackoverflow.com/questions/29475295/javascript-equivalent-for-jquery-on-with-additional-selector This has been asked and answered multiple times because `.on()` is arguably the most powerful feature of jQuery, abstracting binds/listeners away from only existing elements while also improving performance by deduplicating those binds/listeners.
> > > The code already exists, the only question is how to implement it in a wrapper function with methods that anticipate a future native equivalent to `.on()` and when detected switch to leveraging it instead – just like jQuery itself always has.
> > 
> > 
> > I have checked the address you provided, but the on solution you provided is not suitable for use in complex production environments. The biggest problem is that its event cannot call stopPropagation. I think we can still use the on method of jQuery. I have tried customizing jQuery that only includes the on method, and it works well. However, the only problem is that even with code compression, its size is close to 30K . If we can tolerate this size, then we can immediately make select2 dependent on jQuery. In the future, we will gradually search for smaller alternatives to the on method
> 
> jQuery needs to be removed 100%. The "on" method and the way jQuery handles events isn't appropriate for today's web. Javascript can everything jQuery can do, and then some.
> 
> Did you get a chance to see the code I provided above?

Sorry, I replied incorrectly. What did I mean https://stackoverflow.com/questions/29475295/javascript-equivalent-for-jquery-on-with-additional-selector I have checked this address, not the example you provided above

**Comment by RacerDelux** [comment]
@ajiho what would the benefit of that be vs the code I provided above? I don't think an external library is needed here.

**Comment by duanemoody** [comment]
> In fact, the most important issue is to solve the binding problem of events. If we can find a small event handling library similar to jQuery, other refactoring work is not too difficult. In order to solve this problem, I published just on in the comments above to handle the event part. However, just on is actually custom built using jQuery, and its size is still too large, even after compression, it is nearly 30K, which is really difficult to accept. However, now I have a new breakthrough. Its size may only be 1K after being compressed by Gzip, which is an exciting thing. So here I want to tell everyone

Try [this gist](https://gist.github.com/duanemoody/c3f10a0940416895876567e9c5963558). It's the same syntax but refactored down to about 12 loc.

**Comment by RacerDelux** [comment]
> Try [this gist](https://gist.github.com/duanemoody/c3f10a0940416895876567e9c5963558). It's the same syntax but refactored down to about 12 loc.

This kind of goes directly against modern coding standards for JavaScript. At most each dropdown should have a single event listener.

But really, all dropdowns should share a single event listener.

**Comment by duanemoody** [comment]
> This kind of goes directly against modern coding standards for JavaScript. At most each dropdown should have a single event listener.
> 
> But really, all dropdowns should share a single event listener.

I believe that's my point, yes.

**Comment by duanemoody** [comment]
This gist is intended to work the same way as jQuery `.on()`, meaning the event listener is attached once to a shared ancestor element and pointed at a selector, not attached each time a select2 gets created and inserted into DOM. If I misunderstood you my apologies.

**Comment by RacerDelux** [comment]
> This gist is intended to work the same way as jQuery `.on()`, meaning the event listener is attached once to a shared ancestor element and pointed at a selector, not attached each time a select2 gets created and inserted into DOM. If I misunderstood you my apologies.

Gotcha, no that makes sense.
In the past in jQuery though, `.on` was not really used at a global level, but rather for each button or action that one wanted to listen to. While the same philosophy could be used in jQuery, it was not.

Just need to make sure the codebase keeps up with modern programming practices.

**Comment by duanemoody** [comment]
> Gotcha, no that makes sense. In the past in jQuery though, `.on` was not really used at a global level

`.on()` was specifically created for 'global' (rather, below an ancestor that could be `body`) event delegation because its predecessors were inefficient and clogging the event queue. I remember because I was there for all them from `.bind()` onward.

The central purpose behind jQuery was to reproduce IE5's [CSS behaviors](https://learn.microsoft.com/en-us/previous-versions//ms531018(v=vs.85)?redirectedfrom=MSDN), which seamlessly bound events to a selector from page load onward through dynamic element creation, yet John Resig soon realized his event binds only applied to static results at page load, and he/the jQ team struggled with various methods trying to reproduce how IE did it: https://learn.jquery.com/events/history-of-events

Eventually `ancestor.on(eventnames, selector, handler);` nailed it in 2011 with event delegation and has remained the solution since, which is why the gist I posted follows its syntax.

**Comment by kaystrobach** [comment]
Having select2 as webcomponent, encapsulating all the stuff would be a genius first step. after that something like this would then make progressive enhancement and step by step reduction of jQuery possible:

```html
<select2-webwc>
<select ...>...</select>
</select2-webwc>
```

As jQuery would be enclosed in the WC the attack surface is a bit smaller and the objections should be less.
Maybe using the slim version is a good start (have not checked if this works though)

**Comment by stratboy** [comment]
We must say, though, that selectable select elements are finally on the way of becoming a standard w3c feature, therefore I don't know if it's worth to spend time here anymore.

**Comment by RacerDelux** [comment]
> We must say, though, that selectable select elements are finally on the way of becoming a standard w3c feature, therefore I don't know if it's worth to spend time here anymore.

Still totally viable to build a plugin based off the standard select to make it more controllable and easier to use.

**Comment by kaystrobach** [comment]
For sure and select2 has so many awesome stuff inside esp. with the ajax stuff

**Comment by RacerDelux** [comment]
> We just want to simply get rid of jQuery

Already wrote my own select from scratch haha. Couldn't wait.

**Comment by kaystrobach** [comment]
Can be called select3 then. Would be fine for me. What is the maintainer saying about this?

**Comment by RacerDelux** [comment]
> My suggestion is to fork directly in the repository and remove jQuery dependencies while maintaining the original functionality of select2. What do you think?

With the select html element getting better and better support, a rewrite would probably be best.

Plus, all of the function this gives should be able to be done with less than 1000 lines of code really. The one I wrote, while not quite as feature rich, is like 200.

**Comment by RacerDelux** [comment]
> I am leaving this message only to inform those who are concerned about this issue that my event library,[just-event.js](https://github.com/ajiho/just-event.js), has made significant progress. It is now completely implemented independently, with a very small size (gzipped: <= 2KB) and usage similar to jQuery
> 
> [@duanemoody](https://github.com/duanemoody) [@RacerDelux](https://github.com/RacerDelux)
> 
> What do you think?

Honestly, not necessary. While it may make it comfy for some developers who are used to jQuery, I found it better to just learn the new pure JavaScript syntax.

Also, modern coding standards on how even listeners are used are quite different than how they were used in jQuery.

For instance, instead of adding a listener for every button on a form, it is common practice now to use data attributes on each element you wish to trigger on, and a single listener on the parent. You use a switch statement on the data attribute to fall into the correct state.

tldr; Typing `addEventListener` instead of `On` really isn't that big of a change.

**Comment by RacerDelux** [comment]
@ajiho I think you may have forgotten to translate that

I think you have a misunderstanding on the modern approach to JS though.
These days it is considered inappropriate to loop through a list and add an event listener to each `li` element.
You should, instead, listen to the parent `ul`.

Also, you can use CSS to remove user interaction from the inner `<span>` in your example. It allows the button to receive all events.
